subsys/pm/pm.c

```diff
@@ -68,6 +68,17 @@ static inline void pm_state_notify(bool entering_state)
 	k_spin_unlock(&pm_notifier_lock, pm_notifier_key);
 }
 
+static inline int32_t ticks_expiring_sooner(int32_t ticks1, int32_t ticks2)
+{
+	/*
+	 * Ticks are relative numbers that defines the number of ticks
+	 * until the next event.
+	 * Its maximum value is K_TICKS_FOREVER ((uint32_t)-1)
```

Comment:
Now I'm thinking if this function is correct. If the maximum value is max of `uint32_t`, it would overflow both arguments being `int32_t`. Is the error in the comment or in the implementation?

Reply:
@hubertmis I also saw this and was confused.
ticks can be K_TICKS_FOREVER which is either a (uint32_t - 1) or (int64_t - 1), see code below :
```#ifdef CONFIG_TIMEOUT_64BIT
typedef int64_t k_ticks_t;
#else
typedef uint32_t k_ticks_t;
#endif

#define K_TICKS_FOREVER ((k_ticks_t) -1)
```
The next kernel ticks is obtained through this function :
```
_kernel.idle = z_get_next_timeout_expiry();
```
And the implementation of  z_get_next_timeout_expiry will cast the K_TICKS_FOREVER to (int32_t)
See : https://github.com/zephyrproject-rtos/zephyr/blob/main/kernel/timeout.c#L200

So globally the ticks are cast to int32_t even if they can get K_TICKS_FOREVER value
And I think this should be fixed in a separate commit and changed to uint32_t
What do you think ?

Reply:
If I understand correctly, both `ticks` values in this function can have the following values:
1. `-1` meaning FOREVER
2. `0` or a positive number - number of ticks since now

If that's the case, the returned value is incorrect. It does not handle `-1` or other negative values.

Instead, it should be something like:
```
// Validate input parameters
ASSERT(ticks1 >= -1);
ASSERT(ticks2 >= -1);

if (ticks1 == FOREVER) return ticks2;
if (ticks2 == FOREVER) return ticks1;
if (ticks1 < ticks2) return ticks1;
return ticks2;
```

Reply:
You are right, I forgot about the negative numbers.
What do you think about the last implementation ?

Reply:
It looks functionally the same as the one I proposed, just harder to read as an oneliner ;)